### PR TITLE
Java: Add metric queries for counting sinks coming from models

### PR DESCRIPTION
**Pull request by tausbn**
Adds two queries for gathering metrics on the number of alerts (for a selection of queries) that arise from models with the `ai-generated` provenance.

Also adds `pragma[nomagic]` to one of the utility predicates, as it would otherwise cause an inefficient join order to appear.

### CODE REVIEW

File: java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+ */
+string getSinkModelQueryRepr(PotentialSinkModelExpr e) {
+  result = "\nsinkModel: " + getSinkModelRepr(e)
+}
+
+/**
+ * A parameterised module that takes a dataflow config, and exposes a predicate for counting the
+ * number of AI-generated sink models that appear in alerts for that query.
+ */
+private module SinkTallier<DataFlow::ConfigSig Config> {
```
Comment (kaeluka) [inline java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll:144]: a great use of a parameterized module <3

**Review by kaeluka — APPROVED** [review]
I just had a few superficial comments — LGTM, otherwise :)

File: java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+private import semmle.code.java.security.SqlInjectionQuery
+private import semmle.code.java.security.UrlRedirectQuery
+private import semmle.code.java.security.TaintedPathQuery
+private import semmle.code.java.security.SqlInjectionQuery
+private import AutomodelJavaUtil
+
+private newtype TSinkModel =
+  MkSinkModel(
+    string package, string type, boolean subtypes, string name, string signature, string input,
+    string ext, string kind, string provenance
```
[the remark this answers is not in the document]
Comment (tausbn) [reply java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll:17]: This was just a mistake on my part. Thanks for catching it!

File: java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ * A parameterised module that takes a dataflow config, and exposes a predicate for counting the
+ * number of AI-generated sink models that appear in alerts for that query.
+ */
+private module SinkTallier<DataFlow::ConfigSig Config> {
+  module ConfigFlow = TaintTracking::Global<Config>;
+
+  predicate getSinkModelCount(int c, SinkModel s) {
+    s = any(ConfigFlow::PathNode sink).getNode().asExpr().(PotentialSinkModelExpr).getSinkModel() and
+    c =
+      strictcount(ConfigFlow::PathNode sink |
```
[the remark this answers is not in the document]
Comment (tausbn) [reply java/ql/src/Telemetry/AutomodelAlertSinkUtil.qll:150]: Indeed, given that we know what all the sink models are already, the ones with zero results have no useful information and are just taking up space.

**Review by kaeluka — APPROVED** [review]
Thank you!

**Comment by kaeluka** [comment]
@jhelie — I think it was `@tausbn`'s intention to merge this before going on vacation, but may be out of the loop. Do you agree we can merge this?

**Comment by jhelie** [comment]
If those are improvements related to the MRVA metrics queries used in the work you've picked up, then yes I agree let's merge it 👍

**Pull request merged**